### ISSUE TITLE
Terminal can omit backgrounds, links and underlines on empty cells

### ISSUE THREAD

**Issue by PhMajerus** [microsoft/terminal#13229]
**Original Title**: Hyperlink on spaces over-optimization breaks rendering and usability

### Windows Terminal version

1.14.1452.0

### Windows build number

10.0.22000.708

### Other Software

My ActiveScript Shell, just to use JScript to easily script the console, but problem is with VT optimization.

### Steps to reproduce

When a hyperlink contains trailing spaces, they are shown or trimmed depending on the number of trailing spaces.
Starting at 9 spaces or more, they get used as spacing, but are not considered part of the hyperlink, while 8 or less are properly rendered and clickable as part of the hyperlink.

### Expected Behavior

Hyperlinks extending on space characters can be a design decision for consistency and usability, the Terminal should not decide to shorten the number of columns the hyperlink is effectively available on.
Note non-preview build 1.12.10983.0 does not trim the spaces, so this is a new bug introduced in the Preview version 1.14.1452.0.

### Actual Behavior

Current Preview build seems over-agressive in optimizing the spaces, and swallows spaces without abiding by the original VT output of apps.

![image](https://user-images.githubusercontent.com/25664275/172058559-5cc604aa-5779-458b-99b6-9cd922b16287.png)
See how the spaces above the red underlines are different between 8 and 9 spaces.

**Comment by zadjii-msft** [comment]
Hmm. I'm betting we didn't account for hyperlinks in that optimization. Either in https://github.com/microsoft/terminal/blob/main/src/renderer/vt/paint.cpp#L376, or up in Renderer itself.

**Comment by DHowett** [comment]
This probably happens when we decide to use `CUF` instead of `space space space space`:

https://github.com/microsoft/terminal/blob/4e20a8631cbb5b20defe549a0e75bc7719969bee/src/renderer/vt/paint.cpp#L412-L421

However, I don't think we _changed_ this behavior in Preview 1.14...

Thank the gods it's not because of the optimization _I_ added, `HasIdenticalVisualRepresentationForBlankSpace`.

HOWEVER, It would probably be fixed if we DID use `HasIdenticalVisualRepresentationForBlankSpace` here:

https://github.com/microsoft/terminal/blob/4e20a8631cbb5b20defe549a0e75bc7719969bee/src/renderer/vt/paint.cpp#L437

Proving once again that having multiple copies of the same code is deleterious to one's health.

**Comment by DHowett** [comment]
>Note non-preview build 1.12.10983.0 does not trim the spaces

Hmmmmm. What about 1.13?

**Comment by PhMajerus** [comment]
I got the 1.13.11432.0 RTW in the meantime, and it doesn't trim the spaces.

**Comment by j4james** [comment]
FYI, I've been trying to git bisect this, and so far I've narrowed it down to somewhere between fa25dfbf7ab0fac00fc0dbf953ae9d4c5000258a (April 21) and ce850cae184474ecb8a9353afa32d9ae5edaa143 (May 5).

**Comment by zadjii-msft** [comment]
Okay so 
* 1.12: ✅
* 1.13: ✅
* 1.14: ❌


what the devil in 1.14 could have broken this?

**Comment by DHowett** [comment]
The diffs to `paint.cpp` specifically from `release-1.13` to `release-1.14` only cover changing the types from `bool` to `auto` for the locals involved in electing to use spaces or `CUF`.

**Comment by j4james** [comment]
OK. So this is a wider problem than just hyperlinks. You can see the same issue with something like the reverse attribute.

    printf "\e[7m         test         \e[m\n"

Both the spaces before and after the `test` should be reversed, but only the first lot are. This only worked prior to f4e0d9f2bd1f5da3a6c378453b90320fb6b1c542 because conpty was relying on an incorrect implementation of `ECH` in Terminal which wasn't taking into account the rules for erase attributes.

Bottom line is that conpty shouldn't be expecting any of the VT erase operations to write out attributes other than color. This was probably already a problem in other conpty terminals which implement the erase operations correctly.

**Comment by DHowett** [comment]
I wonder, should we simply revisit the ECH/CUF space eliding optimization?

Thanks for digging into this, James :smile:

**Comment by zadjii-msft** [comment]
You know, this makes sense. I did the ECH optimization before #3100, which fixed ECH for conhost. 

> should we simply revisit the ECH/CUF space eliding optimization

I don't think that's a bad idea. I think the worst case of the spaces was "spaces at the end of the line", but I don't think we ECH&CUF for that anyways. We may want to do that for 1.14 Stable, if we can do it safely in the next week.

**Comment by DHowett** [comment]
Right, for excess spaces at the end of the line I think we use `EL`. However, that's *also* the source of some issues...

**Comment by j4james** [comment]
Yeah. Here's a failing test case for `EL`.

    printf "\e[7m"; printf ' %.0s' $(seq 1 $COLUMNS); printf "\e[m\n"

This writes out a full line of spaces with the reverse attribute set, and conpty translates that to an `EL`. That would have worked prior to PR #13024, but won't anymore.

**Comment by DHowett** [comment]
The more I think about it, the closer the resolutions for this bug and #8341 look to eachother. At the end of the day, both are about not printing "meaningful" spaces, where "meaningful" is defined roughly as "spaces that differ from the clear state for new lines or empty buffers." I had a WIP for #8341 going, I should dig it up.

**Comment by DHowett** [comment]
So, here's something I was curious about. We had an issue where we lied about being xterm, and didn't support `REP`. That makes me gun-shy on my next comment:

_Should we consider using `REP` for this optimization?_ Its documented purpose is totally in line with what we need... :D

**Comment by j4james** [comment]
> Should we consider using `REP` for this optimization?

I don't think that's a bad idea. There are definitely terminals that don't support `REP`, but I suspect the kind of terminals that would use conpty would also be fairly likely to support `REP`. Is there a list of known conpty terminals that we can test before we go down this route?

**Comment by zadjii-msft** [comment]
> Is there a list of known conpty terminals that we can test

No, but there probably should be. Off the top of the dome:
* conhost (when used for WSL interop)
* WT (obviously)
* VsCode (by way of xterm.js)
* wezterm (I think)
* sshd.exe on Windows, but that places no limit on the actual terminal on the receiving end.

**Comment by j4james** [comment]
Here are a few more that should work with conpty:

* [Contour](https://github.com/contour-terminal/contour/blob/master/README.md#prerequisites-windows-10-or-newer)
* [Alacritty](https://github.com/alacritty/alacritty/blob/master/README.md#requirements)
* [Mintty](https://github.com/mintty/mintty/blob/master/wiki/Tips.md#inputoutput-interaction-with-alien-programs)

And of those listed so far, I know that Xterm.js, WezTerm, Contour, Alacritty, and Mintty all support `REP` (assuming we aren't getting into any weird edge cases, and just using it to repeat spaces).

But I guess if you're running sshd on Windows, and then logging into it from something like Rxvt or the Linux console (both of which didn't support `REP` last I checked), then you're assumedly going to have problems.

**Comment by PhMajerus** [comment]
I have questions and concerns about this...

I don't know all the rationale for the optimizations done inside ConPTY, but isn't this a bit of a dangerous over-optimization? Especially if it only removes a few bytes from the stream but isn't properly supported through SSH - which is probably where those few saved bytes could even be noticeable at all? Do those optimizations really make a difference when running with a local ConPTY client?

If all terminals were Windows ones using the new ConPTY interface, I would say just publish a page with the minimum features required to properly handle ConPTY output and we'd expect 3rd party terminals to support those... but since we have SSHD, any existing or even legacy terminal could consume that output, wouldn't that mean SSHD would have to parse and re-generate simplified VT output again to ensure it can handle terminals that were not designed according to that minimum features list, disabling the scenarios where those saved bytes would actually travel over more expensive channels than IPC and adding yet another VT parser+renderer to the chain?

Maybe it's better to just make the output of ConPTY follow the same optimizations as done by something already widely used such as (n)curses rendering for an unknown terminal, this would ensure that most terminals that might connect through SSH will support it properly (I don't think ConPTY will have any weight in features supported by terminals on the Un*x side, while ncurses probably does).
Alternatively, maybe ConPTY should default to the most basic output and take advantage of features explicitly supported by the connected terminal, something like an opt-in API function with flags to say which of those optimizations can be used. A terminal that does not call something like `SetPseudoConsoleAllowedFeatures(PSEUDOCONSOLE_OPTINFEATURE_REP | [...other opt-in flags], hPC)` would get normal spaces instead of REPs.

I trust your judgement and actually really appreciate that you're trying to optimize as much as possible, but I'd prefer wasting a few bytes to ensure a predictable behavior than ending up with hard to track rendering issues that will be blamed on Windows CUI apps developers when it really is ConPTY reinterpreting their output. This thing needs to work in all scenarios if we have no control over its optimizations.

**Comment by j4james** [comment]
@PhMajerus Conpty actually does already have a couple of different modes catering for varying levels of VT support, so it is still theoretically possible to support terminals with very basic VT capabilities. I'm not sure if sshd makes use of that functionality though.

**Comment by DHowett** [comment]
Unfortunately, we're likely about to ship Terminal 1.14 broadly with this issue. I'll pin this once it goes out and we can dig into what we should actually do--that is, use REP or spend the bytes or whatever--in the 1.16 cycle.

I'm a bit worried about actually dumping the spaces into the payload, because we've historically had problems with wrapping and emitting characters in the deferred EOL position. By attempting to divine where the "meaingful" end of a line is, and not space up until the end, we're _somewhat_ insulated from that. It's not a good reason, but it is sort-of a reason.

At the end of the day, there's always going to be an issue with ConPTY converting screenfuls of content back into VT. We just have to pick a poison.

**Comment by zadjii-msft** [comment]
Well this is hairy now ain't it. 
* ConPTY has a pile of code in here to deal with how it thought it was gonna use `ECH`. `ECH` doesn't move the cursor. But `REP` would. 
* Obviously we've got lots of tests that actually compared the raw conpty output, so changing that makes the tests sad. 
* Just a delta of `const auto optimalToUseECH = false;` seemingly fixes this in the Terminal, BUT
  * Now, we emit spaces for the row. So we can get failures like
  ```
  ========== Checking the terminal buffer state (after) ==========
  Checking row 0...
  Error: Verify: IsFalse(tb.GetRowByOffset(row).WasWrapForced()) [File: 
  D:\dev\public\terminal\src\cascadia\UnitTests_TerminalCore\ConptyRoundtripTests.cpp, Function: 
  TerminalCoreUnitTests::ConptyRoundtripTests::ResizeInitializeBufferWithDefaultAttrs::<lambda_1>::operator (), Line: 2985]
  EndGroup: TerminalCoreUnitTests::ConptyRoundtripTests::ResizeInitializeBufferWithDefaultAttrs#metadataSet17 [Failed]
  ```
  Because now we're emitting actual spaces. So terminal now thinks the line wrapped.
* Even if we did replace with `REP`, then oh no, I bet the terminal will _still_ think the line wrapped.


I'll need to noodle more (probably next week)

**Comment by j4james** [comment]
It just occurred to me that we may be worrying unnecessarily about the implications of just writing out the spaces. Is it really any different from a line filled with some other character which we wouldn't have been able to optimise? It's not like those spaces could have come from something like an erase screen operation - someone must have actively written out a space in those locations in order for us to get these attributes.

So the idea wouldn't be to turn off the ECH/EL optimisations completely - just turn them off if the spaces are using any of the unsupported attributes.

The only edge case I can think of would possibly be from one of the legacy console APIs that can fill parts of the buffer with attributes like underline and reverse without necessarily writing to those locations. But again I don't think that would be any different from filling the screen with a non-space character, which I think you can also do with the console APIs.

Anyway, it's just a thought, in case you don't have any better ideas.

**Comment by zadjii-msft** [comment]
Shit @j4james I think you and I came to the same idea at the same time. I'm pushing out a PR now 😉
